### DISCUSSION TITLE
Integrating Quick Replies

### DISCUSSION THREAD

**Discussion by jampez77** [tsightler/ring-mqtt#535] [Feature Requests]
I'd like to be able to turn on / off quick replies using this project. Additionally I would like to be able to change the quick reply message based on who is home. 

For instance i would like to set the reply to

-  "Hi, we'll be right there" if someone is home and downstairs
-  "Hi! Please wait. It may take me a moment to answer." if someone is home and _NOT_ downstairs
-  " Please leave the package outside" if we are almost home.

Here is what i've found so far using MITM. Let me know if there is anything else you might need.

This is the endpoint for enabling / disabling the quick replies and setting the response time:

`mode` can be set to `autoreply or disabled` to turn the feature on or off. The response time can be changed by setting the `delay_ms` in the `autoreply_settings` object one of the following values:

- 0 = Right Away
- 2000 = 2 seconds
- 5000 = 5 seconds
- 10000 = 10 seconds
- 15000 = 15 seconds
- 20000 = 20 seconds

PATCH: `https://api.ring.com/devices/v1/devices/96671488/settings`
Payload:
Enabled:
```
{
    "concierge_settings": {
        "alexa_settings": {
            "delay_ms": 10000
        },
        "autoreply_settings": {
            "delay_ms": 0
        },
        "mode": "autoreply"
    }
}
```
DIsabled:
```
{
    "concierge_settings": {
        "alexa_settings": {
            "delay_ms": 10000
        },
        "autoreply_settings": {
            "delay_ms": 0
        },
        "mode": "disabled"
    }
}
```

The specific reply can be set by the calling the following URLs for each reply: 

The payload for all of the following is:
```
{
    "doorbot_ids": [
        96671488
    ],
    "feature": "autoreply"
}
```

  They're at the 30! At the 20! At the 10! Almost to the door! They'll be right there! (Crowd Cheering Background)
  POST: `https://api.ring.com/ars/v1/replies/preset_ff-superbowl_2023-03/default`
  
  [Football Fanfare] If you'd like to leave a message, you can do it now.
  POST: `https://api.ring.com/ars/v1/replies/preset_ff-superbowl_2023-04/default`
  
  Hi, we'll be right there. Feel free to practice your touchdown dance while you wait.
  POST: `https://api.ring.com/ars/v1/replies/preset_ff-superbowl_2023-01/default`
  
  We're watching the game! Leave us a message! (Crowd Cheering Background)
  POST: `https://api.ring.com/ars/v1/replies/preset_ff-superbowl_2023-02/default`
  
  Hi! Please wait. It may take me a moment to answer.
  POST: `https://api.ring.com/ars/v1/replies/preset_06/default`
  
  Sorry, We're not interested.
  POST: `https://api.ring.com/ars/v1/replies/preset_04/default`
  
  Thanks for stopping by. If you'd like to leave us a message, you can do it now.
  POST: `https://api.ring.com/ars/v1/replies/preset_05/default`
  
  We can't answer the door right now but if you'd like to leave a message you can do it now.
  POST: `https://api.ring.com/ars/v1/replies/preset_02/default`
  
  Hi! We'll be right there.
  POST: `https://api.ring.com/ars/v1/replies/preset_03/default`
  
  Please leave the package outside, If you'd like to leave a message, you can do it now.
  POST: `https://api.ring.com/ars/v1/replies/preset_01/default`

I've noticed that after setting the reply a call will be made to this URL which will respond with the reply details for example:
`https://api.ring.com/ars/v1/replies/default?doorbot_id=96671488&feature=autoreply`
```
{
    "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_01.wav",
    "codec": "pcmu",
    "doorbot_id": 96671488,
    "feature": "autoreply",
    "is_default": true,
    "message": "Please leave the package outside. If you’d like to leave a message, you can do it now.",
    "reply_id": "preset_01",
    "reply_type": "preset",
    "updated_at": "2023-01-31T10:50:06.888Z"
}
```

Additionally you can query the full list of available replies here:

GET: `https://api.ring.com/ars/v1/replies?doorbot_id=96671488&feature=autoreply`
Response:
```
{
    "codec": "pcmu",
    "feature": "autoreply",
    "replies": [
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_ff-superbowl_2023-03.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "They’re at the 30! At the 20! At the 10! Almost to the door! They’ll be right there! (Crowd Cheering Background)",
            "reply_id": "preset_ff-superbowl_2023-03",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_ff-superbowl_2023-04.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "[Football Fanfare] If you’d like to leave a message, you can do it now.",
            "reply_id": "preset_ff-superbowl_2023-04",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_ff-superbowl_2023-01.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Hi, we’ll be right there. Feel free to practice your touchdown dance while you wait.",
            "reply_id": "preset_ff-superbowl_2023-01",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_ff-superbowl_2023-02.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "We’re watching the game! Leave us a message! (Crowd Cheering Background)",
            "reply_id": "preset_ff-superbowl_2023-02",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_06.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Hi! Please wait. It may take me a moment to answer.",
            "reply_id": "preset_06",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_04.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Sorry, we're not interested.",
            "reply_id": "preset_04",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_05.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Thanks for stopping by. If you’d like to leave a message, you can do it now.",
            "reply_id": "preset_05",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_02.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "We can’t answer the door right now, but if you’d like to leave a message, you can do it now.",
            "reply_id": "preset_02",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_03.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Hi! We’ll be right there.",
            "reply_id": "preset_03",
            "reply_type": "preset"
        },
        {
            "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_01.wav",
            "codec": "pcmu",
            "doorbot_id": 96671488,
            "feature": "autoreply",
            "message": "Please leave the package outside. If you’d like to leave a message, you can do it now.",
            "reply_id": "preset_01",
            "reply_type": "preset"
        }
    ]
}
```

You can also query the reply currently set:
GET: `https://api.ring.com/ars/v1/replies/default?doorbot_id=96671488&feature=autoreply`
Response:
```
{
    "audio_url": "https://dmgxmf8evaeu9.cloudfront.net/presets/preset_01.wav",
    "codec": "pcmu",
    "doorbot_id": 96671488,
    "feature": "autoreply",
    "is_default": true,
    "message": "Please leave the package outside. If you’d like to leave a message, you can do it now.",
    "reply_id": "preset_01",
    "reply_type": "preset",
    "updated_at": "2023-01-31T10:50:06.888Z"
}
```

**Comment by tsightler** [comment]
It's interesting information, but I don't see this as something I'm likely to implement in the near future.  There's definitely lots of questions as to how it could be presented and selected reasonably via MQTT.  I feel like a user would want to see at least the majority of the message they are selecting, but that makes the actual selection via MQTT awkward since, as far as I know, HA MQTT Select requires that the command match the option selected exactly.  This would technically work, it's just ugly, IMO.

Regardless, I just don't have much time to play with this right now and I think it's a pretty niche request so, unless you want to code it up and send through a PR for review, it's not going to be something I'll be able to look at anytime soon.

**Reply by jampez77** [reply]
That's fair enough. I guess it's pretty niche and I thought it might not be simple to implement.

If I find the time I will look into attempting some code but my Python skills are pretty basic so I'm not sure how likely that will be.

I suppose at the very least if they allow custom replies at some point this could be a baseline to build from.

**Reply by tsightler** [reply]
I mean, I find it somewhat interesting, and it wouldn't be _**that**_ hard to implement overall, but I just have to be realistic about what I can support given limited time and, honestly, interest.  Every device, and every feature is something that I have to write, test, support, worry about when it will break, etc.

At this point the Ring ecosystem has grown quite large since I first started this project, and they just keep introducing more and more device that I don't own and features that I mostly don't care about, yet every one of those features is something that people will ask me to support and open issues when it breaks or doesn't work how they want it too.

So I'm not saying I will never do it, if more people come along and ask for this feature, it might move up in priority, but for now, it will be a low priority enhancement request which I'll only likely get to if I'm bored at some point.

BTW, this project is 99% Javascript, the other 1% is just Bash shell, and there is no Python at all.

**Reply by jampez77** [reply]
Ah fair enough. I've put it on my own to do pile as well. It's been a long time since i've properly delved into JS but always happy and excited to (re)learn.

**Reply by xPudda** [reply]
Hi everyone! I was trying to understand if there was the possibility to do these but unfortunately it is not available yet. 
Are there any ideas for implementing it? 
That would be really helpful! It would be also great just with default presets.

Overall, thanks for your hard work @tsightler!

**Reply by tsightler** [reply]
I actually find the idea very interesting, but I just don't personally have a use for it, and thus finding the time to implement it, test it, etc just isn't there for me right now.  Ideally, I'd need to get some functions pushed upstream to ring-client-api.

Pretty much all of the information required is above.  It would require identifying devices that actually support this, etc.  I'm happy to entertain PRs, but I just don't think I'll be doing it myself any time soon.

One thing I have considered, is a generic MQTT-to-DeviceAPI proxy endpoint, this would allow an MQTT user to send any known commands to the device without requiring specific support in ring-mqtt itself.  Obviously useful for advanced users only, but it would allow users to build complex automations for unsupported features like this.

**Reply by xPudda** [reply]
I would be happy to test it if needed ;)
